basic_pipeline/4_Caps.md

```diff
@@ -0,0 +1,116 @@
+
+An element can send caps as one of the [actions](https://hexdocs.pm/membrane_core/Membrane.Element.Action.html) it can take - the [`:caps` action](https://hexdocs.pm/membrane_core/Membrane.Element.Action.html#t:caps_t/0).
+
+Another thing is that we can specify the behavior of an element when it receives the caps with the use of [`handle_caps/4` callback](https://hexdocs.pm/membrane_core/Membrane.Element.WithInputPads.html#c:handle_caps/4).
+
+For all the filter elements, `handle_caps/4` has a default implementation, which is relaying the received caps on all the output pads of that filter.
+However, if your filter is changing the format of data being sent, it should override the implementation of that callback to prevent caps flying through it, and send the proper caps via the output pads. 
+
+For the source element, it is necessary to send the caps as in each pipeline the source is the first element - caps wouldn't flow through the pipeline if the source element wouldn't have sent them. Sending can be done in the `handle_stopped_to_prepared/2` callback.
+# Example
+Imagine a pipeline, which starts with the source producing a video, which is then passed to the filter, responsible for reducing the quality of that video if it is too high.
+For the source element, we should have the `:output` pads caps which would allow us to send video in the higher and in the lower quality. The same caps should be specified on the input of the filter element. However, the caps on the output of the filter should accept only video in the lower quality.
+Here is the definition of the source element:
+```Elixir
+# Source element
+
+defmodule Source do
+ def_output_pad(:output, 
+   demand_unit: :buffers, 
+   caps: [
+      {Format.Raw, pixel_format: one_of([:I420, :I422]), framerate: range(30, 60), width: 480, height: 300},
+      {Format.Raw, pixel_format: one_of([:I420, :I422]), framerate: range(30, 60), width: 720, height: 480}
+   ]
+ )
+ ...
+ def handle_stopped_to_prepared(_ctx, state) do
+ ...
+   {{:ok, [caps: {:output, %Formats.Raw{pixel_format: I420, framerate: 45, width: 720, height: 300}}]}, state}
+ end
+```
+While returning from the `handle_stopped_to_prepared/2` callback, the element will send the caps described by the `Formats.Raw` structure, through the `:output` pad.
+Will those caps meet the caps specification provided by us? Think about it!
+In fact, they will. The format matches (both in the caps being sent and in the caps specification of the pad, we have `Format.Raw` module). When it comes to the options, we see, that `I420` is in the `one_of` list, acceptable by the caps specification format for `width` equal to 720 and `height` equal to 480, and the `framerate`, equal to 45, is in the `range` between 30 and 45, as defined in the caps specification.
```

Comment:
I believe you meant that 45 is in range between 30 and 60.

Reply:
done